tracer/src/Datadog.Trace/Util/SamplingHelpers.cs

```diff
@@ -32,8 +28,20 @@ internal static bool SampleByRate(TraceId traceId, double rate) =>
         /// <param name="id">The 64-bit id of the object. For example, a span id.</param>
         /// <param name="rate">The sampling rate to apply.</param>
         /// <returns><c>true</c> if the object should be sampled (kept), <c>false</c> otherwise.</returns>
-        internal static bool SampleByRate(ulong id, double rate) =>
-            ((id * KnuthFactor) % Modulo) <= (rate * Modulo);
+        internal static bool SampleByRate(ulong id, double rate)
+        {
+            if (rate == 1.0)
```

Comment:
You generally don't want to compare doubles directly with `==` in C#.

> The `Equals` method should be used with caution, because two apparently equivalent values can be unequal due to the differing precision of the two values.

- https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-double-equals#precision-in-comparisons
- https://www.jetbrains.com/help/rider/CompareOfFloatsByEqualityOperator.html

Consider instead something like:
```suggestion
            if (Math.Abs(rate - 1.0) < TOLERANCE)
```

Reply:
Good question. An alternative could be to use this? (which is also more defensive, in case `rate` it not between 0 and 1)
```csharp
if (rate > 0.999999) // how much precision?
```

Reply:
On second though, precision is probably not an issue with `0.0` and `1.0` specifically, since the fractional part is zero. So let's just leave it as-is 😅